modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C

```diff
@@ -77,58 +77,90 @@ ComputeReturnMappingStress::updateQpStress(RankTwoTensor & strain_increment,
       << std::endl;
   }
 
-  // compute trial stress
-  RankTwoTensor stress_last_iteration = _elasticity_tensor[_qp] * (strain_increment + _elastic_strain_old[_qp]);
-
   RankTwoTensor inelastic_strain_increment, elastic_strain_increment;
   inelastic_strain_increment.zero();
 
-  Real equivalent_delta_stress, first_equivalent_delta_stress;
+  Real l2norm_delta_stress, first_l2norm_delta_stress;
   unsigned int counter = 0;
 
+  std::vector<RankTwoTensor> inelastic_strain_increment_new;
+  inelastic_strain_increment_new.resize(_models.size());
+
+  for (unsigned i_rmm = 0; i_rmm < _models.size(); ++i_rmm)
+    inelastic_strain_increment_new[i_rmm].zero();
+
+  RankTwoTensor stress_max, stress_min;
+
   do
   {
-    elastic_strain_increment = strain_increment;
-    stress_new = _elasticity_tensor[_qp] * (elastic_strain_increment - inelastic_strain_increment + _elastic_strain_old[_qp]);
-
-    for (unsigned i_rmm =0; i_rmm < _models.size(); ++i_rmm)
+    for (unsigned i_rmm = 0; i_rmm < _models.size(); ++i_rmm)
     {
       _models[i_rmm]->setQp(_qp);
+
+      elastic_strain_increment = strain_increment;
+
+      for (unsigned j_rmm = 0; j_rmm < _models.size(); ++j_rmm)
+        if (i_rmm != j_rmm)
+          elastic_strain_increment -= inelastic_strain_increment_new[j_rmm];
+
+      stress_new = _elasticity_tensor[_qp] * (elastic_strain_increment + _elastic_strain_old[_qp]);
+
       _models[i_rmm]->updateStress(elastic_strain_increment,
-                                    inelastic_strain_increment,
-                                    stress_new);
+                                   inelastic_strain_increment,
+                                   stress_new);
+
+      inelastic_strain_increment_new[i_rmm] = inelastic_strain_increment;
+
+      if (i_rmm == 0)
+      {
+        stress_max = stress_new;
+        stress_min = stress_new;
+      }
+      else
+      {
+        for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+        {
+          for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+          {
+            if (stress_new(i,j) > stress_max(i,j))
+              stress_max(i,j) = stress_new(i,j);
+
+            if (stress_min(i,j) > stress_new(i,j))
```

Comment:
Also, maybe it's just me, but I think removing the blank line above this would improve readability.